src/serializer/modules.js

```diff
@@ -75,18 +79,47 @@ class ModuleTracer extends Tracer {
     performCall: () => Value
   ): void | Value {
     let realm = this.modules.realm;
-    if (F === this.modules.getRequire() && this.modules.delayUnsupportedRequires && argumentsList.length === 1) {
+    if (
+      F === this.modules.getRequire() &&
+      !this.modules.disallowDelayingRequiresOverride &&
+      argumentsList.length === 1
+    ) {
       let moduleId = argumentsList[0];
       let moduleIdValue;
       if (moduleId instanceof NumberValue || moduleId instanceof StringValue) {
         moduleIdValue = moduleId.value;
-        if (!this.modules.moduleIds.has(moduleIdValue)) {
+        if (!this.modules.moduleIds.has(moduleIdValue) && this.modules.delayUnsupportedRequires) {
           this.modules.logger.logError(moduleId, "Module referenced by require call has not been defined.");
         }
       } else {
-        this.modules.logger.logError(moduleId, "First argument to require function is not a number or string value.");
+        if (this.modules.delayUnsupportedRequires) {
+          this.modules.logger.logError(moduleId, "First argument to require function is not a number or string value.");
+        }
+        return undefined;
+      }
+
+      // Even if we're not delaying unsupported requires, we want to record what
+      // modules have been initialized
+      if (!this.modules.delayUnsupportedRequires) {
+        if (
+          (this.requireStack.length === 0 || this.requireStack[this.requireStack.length - 1] !== moduleIdValue) &&
+          !this.modules.initializedModules.has(moduleIdValue) &&
```

Comment:
Instead of checking `!this.modules.initializedModules.has(moduleIdValue)` I'd rather see an invariant that if there is already a value, then it better be the same.